### ISSUE TITLE
Introducing new tool for evaluation - Codacy

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#1850]
Github is making it easier to find CI tools, hurray! One is Codacy, free for open source: https://www.codacy.com

In summary it looks to be a web version of checkstyle. Looks like it would provide common web views to view checkstyle issues and give us reporting.

I clicked the integrate button and that seems to be all that was needed for a basic setup. I'm interested in evaluating this tool a bit more, please provide feedback positive or negative here. Based on that we'll see if we want to run with this long term or not

As a bit of a walk through, a few things look interesting:
- web app to look at checkstyle results:
![checkstyle_results](https://user-images.githubusercontent.com/12397753/26993267-3f711a42-4d17-11e7-9d5b-7e0e65149c35.png)

- Runs additional linting tools for us:
![check_tools](https://user-images.githubusercontent.com/12397753/26993298-57c6f1d4-4d17-11e7-97c5-db5ab6f7c589.png)

- Some dashboarding for trending:
![dashboard](https://user-images.githubusercontent.com/12397753/26993305-5ffb3626-4d17-11e7-9d75-606f051988c5.png)

- Nice summary view of PRs:
![pr_view](https://user-images.githubusercontent.com/12397753/26993316-6a26e9a6-4d17-11e7-867b-0132c5a6a2b2.png)

- Trying it out, seems there is not much for project integration
![pr_integration](https://user-images.githubusercontent.com/12397753/26993336-7c9ad1c4-4d17-11e7-91fc-f29e9e13cd41.png)

If this works out we can probably suppress the checkstyle output in travis. This does not replace the checkstyle we have, but IMO the addition of reporting is useful.

**Comment by ron-murhammer** [comment]
Seems potentially useful. I haven't ever tried it before but probably worth giving a shot.

**Comment by DanVanAtta** [comment]
This is pretty impressive : )

![impressive](https://user-images.githubusercontent.com/12397753/26998464-d35feec4-4d38-11e7-96d6-7c2d1af6c659.png)

**Comment by ssoloff** [comment]
Looks good, @DanVanAtta.  Agree that this kind of reporting is a nice step forward.

The only thing I'd suggest investigating further is that, if we enable a tool in Codacy, we should ensure we keep the tool's configuration in source control so that those of us without admin dashboard access can easily see what the rules are and possibly run the tool locally, if needed.  From your screenshots above, it looks like Codacy does this with Checkstyle, so I'm assuming it's possible with most of the other tools, as well.

**Comment by ssoloff** [comment]
@DanVanAtta Following up on my previous comment...  Can you please export the PMD configuration Codacy is using?  I would like to be able to run PMD prior to submitting changes to ensure I don't introduce new violations.  I enabled the PMD Gradle plugin locally, but it's default configuration appears to be significantly different from Codacy's.

Also, there seems to be some differences in code style rules between Checkstyle and PMD.  We probably want to modify the PMD configuration to match the Checkstyle code style rules (or vice versa).  It looks like Codacy will [pick up](https://support.codacy.com/hc/en-us/articles/207994335-Code-Patterns) a customized PMD configuration in source control if one is present.

**Comment by DanVanAtta** [comment]
I'm happy we've been receptive to new tools like Codacy. I understand a bit better your point @ssoloff that we should be able to run the new PMD rules locally before submitting a PR.


Realizing now that we are introducing PMD, checking PMD out I've a few concerns about it:
- tooling does not look as nicely integrated as compared to checkstyle (at least looking at the intellij plugins, the checkstyle one is nice, the PMD one gets a low review rating and is basic / barely usable)
- seems like PMD encourages direct editing of an XML for rules.. I'm guessing there is a rule editor somewhere I missed, but so far surprised that seems to be modus operandi.
- we introduce an overlap problem with checkstyle
- we incur the same "problems" as with using checkstyle, but now with 2 different tools instead of 1, there is overhead to using linting tools. Our goal is good code quality and to automate things that would otherwise come up in a code review. The extra rules we gain may not be worth the overhead.

I think it's a pretty solid conclusion if we are going to use PMD, as @ssoloff  has recommended we would want to use a rules file with Codacy. For now I"ve disabled the PMD check in Codacy, there does not look to be an easy way to export a XML ruleset file. I'm convincing myself that I"m not sold on PMD. We can revisit the tool independent of Codacy (and introducea  rule file and turn on Codacy to pick it up if we choose to do so).

**Comment by ssoloff** [comment]
Sounds good, @DanVanAtta.

I've crafted a custom PMD ruleset before, and it definitely is more involved than starting from a base Checkstyle configuration and tweaking it.  What I found works best is to let Checkstyle be the code style authority, and then remove (rather than duplicate) all code style rules from the PMD configuration.  IMO, let Checkstyle and PMD do what they each do best: syntax analysis and static analysis, respectively.
